docs/sphinx/user_guide/feature/resource.rst

```diff
@@ -0,0 +1,268 @@
+                       to a destination location from the
+                       resource's backend.
+ memset                Set memory value per the resourse's
+                       given backend.
+ wait_for              Enqueue a wait on the resource's stream/thread
+                       for a user passed event to occur.
+ ===================== ===============================================
+  
+Each resource type also defines specific backend information/functionality. For example, each
+Cuda resource contains a ``cudaStream_t`` value with an associated get method. See the 
+individual functionality for each resource in ``raja/tpl/camp/include/resource/``.
+
+
+------------
+Type-Erasure
+------------
+
+Resources can be declared in two formats: An erased resource, and a concrete resource. The 
+underlying runtime functionality is the same for both formats. An erased resource allows a user 
+the ability to change the resource backend at runtime. 
+
+Concrete Cuda resource::
+
+    RAJA::resources::Cuda my_cuda_res;
+
+Erased resource::
+
+    if (use_gpu)
+      RAJA::resources::Resource my_res{RAJA::resources::Cuda()};
+    else
+      RAJA::resources::Resource my_res{RAJA::resources::Host()};
+
+
+Memory allocation on resources::
+
+    int* a1 = my_cuda_res.allocate<int>(ARRAY_SIZE);
+    int* a2 = my_res.allocate<int>(ARRAY_SIZE);
+
+If ``use_gpu`` is ``true``, the underlying type of ``my_res`` is a Cuda resource, therefore ``a1`` 
+and ``a2`` will both be allocated on the GPU. If ``use_gpu`` is ``false``, then only ``a1`` is 
+allocated on the GPU, and ``a2`` is allocated on the host.
+
+
+------
+Forall
+------
+
+A resource is an optional argument to a ``RAJA::forall`` call. It is passed as the first argument 
+when the forall is templated on the execution policy. When the execution policy is passed by value,
+the resource is passed as the second argument. This maintains the order of inputs that can be 
+passed to a ``RAJA::forall`` call::
+
+    RAJA::forall<ExecPol>(my_cuda_res, .... )
+
+or::
+
+    RAJA::forall(ExecPol(), my_cuda_res, .... )
+
+
+When specifying a Cuda/Hip resource the ``RAJA::forall`` is executed aynchronously on a stream.
+Currently Cuda/Hip are the only Resources that enable asynchronous threading with a ``RAJA::forall``.
+All other calls are defaulted to using the ``Host`` resource until further support is 
+added.
+
+The Resource type that is passed to a ``RAJA::forall`` call must be a concrete type. This is to
+allow compile time assertion that the resource is compatible with the given execution policy. For
+example::
+    
+    using ExecPol = RAJA::cuda_exec_async<BLOCK_SIZE>;
+    RAJA::resources::Cuda my_cuda_res;
+    RAJA::resources::Resource my_res{RAJA::resources::Cuda()};
+    RAJA::resources::Host my_host_res;
+
+    RAJA::forall<ExecPol>(my_cuda_res, .... ) // Compiles.
+    RAJA::forall<ExecPol>(my_res, .... ) // Compilation Error. Not Concrete.
```

Comment:
Is there not an underlying resource type trait or something that would allow the second call to work?

Reply:
No, we had decided that we weren't gong to allow users to pass erased Resource types to forall. This was to enable compile time checking of the resource type against the execution policy.

Reply:
Gotcha. Thanks.

Reply:
As a small followon here, for cases where the function is designed to be able to handle dynamic switching it might make sense to allow it.  @artv3's teams loop stuff is a good example, but since the policy has to be concrete for these making it checkable seemed the right way to go for forall at least until there's something where the policy is polymorphic or runtime selected.

If the user has a dynamic resource and they want to use it in a call like this, they should use the `.get()` or `.try_get()` depending on their feelings on exceptions:

```c++
RAJA::forall<ExecPol>(my_res.get<camp::resources::Cuda>(),...);
```

Reply:
Thanks @trws